samples/nrf_desktop/src/modules/config.c

```diff
@@ -0,0 +1,221 @@
+		if (loaded_data[id]) {
+
+			struct config_event *event = new_config_event();
+			event->id = id;
+			event->stored = true;
+			memcpy(event->data, loaded_data[id],
+			       sizeof(event->data));
+
+			EVENT_SUBMIT(event);
+
+			k_free(loaded_data[id]);
+		}
+	}
+
+	return 0;
+}
+
+static int update_config(const u8_t config_id, u8_t *data)
+{
+	int err = 0;
+
+	char key[KEY_LEN];
+	snprintk(key, sizeof(key), "%s/%u", MODULE_NAME, config_id);
+
+	LOG_INF("Store %s", log_strdup(key));
+	err = settings_save_one(key, data, REPORT_SIZE_USER_CONFIG - 1);
+
+	return err;
+}
+
+static int enable_settings(void)
+{
+	int err = 0;
+
+	if (IS_ENABLED(CONFIG_SETTINGS)) {
+		if (IS_ENABLED(CONFIG_DESKTOP_CONFIG_CHANNEL_ENABLE)) {
+			static struct settings_handler sh = {
+				.name = MODULE_NAME,
+				.h_set = settings_set,
+				.h_commit = commit,
+			};
+
+			err = settings_register(&sh);
+			if (err) {
+				LOG_ERR("Cannot register settings handler");
+				goto error;
+			}
+		}
+
+		err = settings_load();
+		if (err) {
+			LOG_ERR("Cannot load settings");
+			goto error;
+		}
+
+		LOG_INF("Settings loaded");
+	}
+
+error:
+	return err;
+}
+
+static void init(void)
+{
+	int err = enable_settings();
+
+	if (err) {
+		module_set_state(MODULE_STATE_ERROR);
+	} else {
+		module_set_state(MODULE_STATE_READY);
+	}
+}
+
+static bool event_handler(const struct event_header *eh)
+{
+	if (IS_ENABLED(CONFIG_DESKTOP_CONFIG_CHANNEL_ENABLE)) {
+		if (is_config_event(eh)) {
+			struct config_event *event = cast_config_event(eh);
+
+			if (!event->stored) {
+				int err = update_config(event->id, event->data);
+
+				if (err) {
+					LOG_ERR("Cannot store value, err %d",
+						err);
+				}
+			}
+
+			return false;
+		}
+	}
+
+	if (is_module_state_event(eh)) {
+		const void * const req_modules[] = {
+#if CONFIG_DESKTOP_HIDS_ENABLE
+			MODULE_ID(hids),
+#endif
+#if CONFIG_DESKTOP_BAS_ENABLE
+			MODULE_ID(bas),
+#endif
+#if CONFIG_DESKTOP_BLE_ADVERTISING_ENABLE
+			MODULE_ID(ble_adv),
+#endif
+#if CONFIG_DESKTOP_MOTION_OPTICAL_ENABLE
+			MODULE_ID(motion),
+#endif
+		};
+
+		struct module_state_event *event = cast_module_state_event(eh);
+
+		/* Settings need to be loaded after GATT services are setup,
+		 * otherwise the values stored in flash will not be written
+		 * to GATT database.
+		 */
+		static unsigned int req_state;
+
+		for (size_t i = 0; i < ARRAY_SIZE(req_modules); i++) {
+			if (check_state(event, req_modules[i], MODULE_STATE_READY)) {
+				unsigned int flag = 1UL << i;
+
+				/* Catch double initialization of any module */
+				__ASSERT_NO_MSG((req_state & flag) == 0);
+
+				req_state |= flag;
+
+				if (req_state == BIT_MASK(ARRAY_SIZE(req_modules))) {
+					init();
+				}
+
+				break;
+			}
+		}
+
+		return false;
+	}
+
+	/* If event is unhandled, unsubscribe. */
+	__ASSERT_NO_MSG(false);
+
+	return false;
+}
+
+EVENT_LISTENER(MODULE, event_handler);
+EVENT_SUBSCRIBE(MODULE, module_state_event);
+EVENT_SUBSCRIBE(MODULE, config_event);
```

Comment:
subscriber only if `CONFIG_DESKTOP_CONFIG_CHANNEL_ENABLE`
we could create some helper for that in event manager header... (not now)

Reply:
done